# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1705833932 -3600
# Parent  5293c7b279807f8f7f8d9a7ec7b25552c66e62be
NNN7 - Fix wrong folder selection at startup.

When restoring the first tab, sometimes the top-most folder gets selected before
later the right folder is selected. Waiting before the "select" event dispatch
seems to fix this issue.

diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -91,16 +91,18 @@ var messageBrowser;
  */
 var multiMessageBrowser;
 /**
  * A XUL browser that displays Account Central when an account's root folder
  * is selected.
  */
 var accountCentralBrowser;
 
+var initialDisplayComplete = false;
+
 window.addEventListener("DOMContentLoaded", async event => {
   if (event.target != document) {
     return;
   }
 
   UIDensity.registerWindow(window);
   UIFontSize.registerWindow(window);
 
@@ -124,16 +126,45 @@ window.addEventListener("DOMContentLoade
     console.warn(`Couldn't restore state: ${e.message}`, e);
   }
   delete window.openingState;
 
   // Finally, add the folderTree listener and trigger it. Earlier events
   // (triggered by `folderPane.init` and possibly `restoreState`) are ignored
   // to avoid unnecessarily loading the thread tree or Account Central.
   folderTree.addEventListener("select", folderPane);
+
+  // In some cases `folderTree.selectedIndex` isn't set correctly when the event is
+  // handled, so the wrong folder is selected at first. We've seen
+  // index 0, so in a unified view, that will be the unified inbox as the top-most row.
+  // If that's grouped-by-sort, then long delays are expected.
+  // Fixing the long delays in https://bugzilla.mozilla.org/show_bug.cgi?id=1875577
+  // will paper over the issue some more.
+  // Sadly folderPane.init() initializes the index to 0, so if it's 0 here after
+  // `restoreState()`, either 0 was indeed the desired the selection or it's not
+  // correctly initialized yet. `restoreState()` looks like a sync function, but
+  // `initialDisplayComplete` isn't always `true` here. Note that `displayFolder()`
+  // sets the index to what is required.
+  let count = 0;
+  let timeWaited = 0;
+  while (count++ < 5 && !initialDisplayComplete) {
+    timeWaited += 100;
+    await new Promise(resolve => setTimeout(resolve, 100));
+  }
+  if (timeWaited) {
+    if (initialDisplayComplete) {
+      console.info(
+        `3pane init: Waited ${timeWaited} ms and index of selected folder is ${folderTree.selectedIndex}`
+      );
+    } else {
+      console.warn(
+        `3pane init: Waited ${timeWaited} ms and index of selected folder is ${folderTree.selectedIndex}`
+      );
+    }
+  }
   folderTree.dispatchEvent(new CustomEvent("select"));
 
   // Attach the progress listener for the webBrowser. For the messageBrowser this
   // happens in the "aboutMessageLoaded" event from aboutMessage.js.
   // For the webBrowser, we can do it here directly.
   top.contentProgress.addProgressListenerToBrowser(webBrowser);
 
   mailContextMenu.init();
@@ -6589,17 +6620,19 @@ function restoreState({
         "messagepaneboxwrapper",
         "collapsed"
       ) !== "true";
   }
   paneLayout.messagePaneSplitter.isCollapsed = !messagePaneVisible;
 
   if (folderURI) {
     displayFolder(folderURI);
+    initialDisplayComplete = true;
   } else if (syntheticView) {
+    initialDisplayComplete = true;
     // In a synthetic view check if we have a previously edited column layout to
     // restore.
     if ("getPersistedSetting" in syntheticView) {
       let columnsState = syntheticView.getPersistedSetting("columns");
       if (!columnsState) {
         threadPane.restoreDefaultColumns();
         return;
       }
